DOC: update the DataFrame.stack docstring

samuelsinayoko opened the pull request:
Checklist for the pandas documentation sprint (ignore this if you are doing
an unrelated PR):

- [X] PR title is "DOC: update the DataFrame.stack docstring"
- [X] The validation script passes: `scripts/validate_docstrings.py <your-function-or-method>`
- [X] The PEP8 style check passes: `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] The html version looks good: `python doc/make.py --single <your-function-or-method>`
- [X] It has been proofread on language by another sprint participant

Please include the output of the validation script below between the "```" ticks:

```
# paste output of "scripts/validate_docstrings.py <your-function-or-method>" here
# between the "```" (remove this comment, but keep the "```")

```

If the validation script still gives errors, but you think there is a good reason
to deviate in this case (and there are certainly such cases), please state this
explicitly.


Checklist for other PRs (remove this part if you are doing a PR for the pandas documentation sprint):

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

samuelsinayoko committed:
Fix docstring or pandas.DataFrame.stack.

- Make description and summary clearer.
- Fix doctests

samuelsinayoko committed:
Polish the docstring (plural issues and the like).

samuelsinayoko committed:
Add description to example.

samuelsinayoko committed:
Add an example with multi-level column.

samuelsinayoko committed:
Add more examples.

samuelsinayoko committed:
Fix sphinx docs

samuelsinayoko committed:
Fix parameter types

samuelsinayoko committed:
Post review improvements.

Reviewed by Marco.

samuelsinayoko committed:
Start refactoring the examples.

Separate creation of data with examples.

samuelsinayoko committed:
Refactor examples

samuelsinayoko committed:
Polish examples.

samuelsinayoko committed:
Add an example where multiple levels are stacked at once.

samuelsinayoko committed:
Clarify filling behaviour with missing values

samuelsinayoko committed:
flake8

datapythonista commented on pandas/core/frame.py:
```diff
@@ -5145,36 +5145,151 @@ def pivot_table(self, values=None, index=None, columns=None,
+             onto column axis.
+        DataFrame.pivot: reshape dataframe from long format to wide
+             format.
+        DataFrame.pivot_table: create a spreadsheet-style pivot table
```
It should be a space between the colon, and the description should start with a capital letter.

samuelsinayoko replied:
Fixed in 652f7b2

datapythonista commented on pandas/core/frame.py:
```diff
@@ -5145,36 +5145,151 @@ def pivot_table(self, values=None, index=None, columns=None,
+             b    1
+        two  a    2
+             b    3
+        dtype: int64
```
It's just a personal opinion, but I think defining all the data first with descriptive names make it a bit more complex to understand.

We could have separate sections for each case, with a title in bold (surrounding the text with double stars, followed by the data creation, using simply `df` in all the cases.

Also, in this case I think it would make the example easier to understand using more real-world examples. As `a`, `b`... don't have a meaning, it'd a bit harder to understand what's going on.

A minor thing, when creating the data, I think it makes more sense that each row is defined as a tuple, than as a list.

For example:

```
**Single level**

>>> df = pd.DataFrame([(8, 12), (22, 35)],
...                   index=['cat', 'dog'],
...                   columns=['weight', 'max_speed'])
>>> df

>>> df.stack()
```

samuelsinayoko replied:
:+1: fixed in 15902ed

datapythonista reviewed (commented):
Great work. Added some comments about formatting, and some that in my opinion should make the examples better.

samuelsinayoko committed:
Put Examples section at the end.

samuelsinayoko commented on pandas/core/frame.py:
```diff
@@ -5145,36 +5145,151 @@ def pivot_table(self, values=None, index=None, columns=None,
+        (levels) being re-organised from a horizontal position (column
+        levels) to a vertical position (index levels).
 
         Examples
```
Fixed in 99734ac

samuelsinayoko committed:
Fix 'See Also' section.

samuelsinayoko committed:
Create separate section for single level columns.

More clear than lumping all the definitions into a single section at the start.

samuelsinayoko committed:
Split the examples into several sections.

Easier to follow.

samuelsinayoko committed:
remove unwanted blank lines

samuelsinayoko committed:
Start using more meaningful index & column names

samuelsinayoko committed:
Use more meaningful column and index names.

pep8speaks commented:
Hello @samuelsinayoko! Thanks for updating the PR.

 - In the file [`pandas/core/frame.py`](https://github.com/pandas-dev/pandas/blob/718f212f319368202e1fe3c3f6117dde27d7b435/pandas/core/frame.py), following are the PEP8 issues :

> [Line 5219:80](https://github.com/pandas-dev/pandas/blob/718f212f319368202e1fe3c3f6117dde27d7b435/pandas/core/frame.py#L5219): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 5240:80](https://github.com/pandas-dev/pandas/blob/718f212f319368202e1fe3c3f6117dde27d7b435/pandas/core/frame.py#L5240): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)

TomAugspurger commented on pandas/core/frame.py:
```diff
@@ -5145,36 +5145,163 @@ def pivot_table(self, values=None, index=None, columns=None,
-        column labels) having a hierarchical index with a new inner-most level
-        of row labels.
-        The level involved will automatically get sorted.
+        Stack the prescribed level(s) from the column axis onto the index
```
This should be a single line. Can you shorten by "column axis" -> "columns" and "index axis" -> "index"?

samuelsinayoko replied:
Thanks for the review. Fixed in a2c9b1a. 

I've also modified the description in the Notes section. It was never completely clear to me why method was called stack (I think I was imagining the column as a board being moved from an horizontal position to a vertical position, whereas I think the name comes from a collection of items being moved from a side by side position to a stack), so I've tried to explain that in the notes section. Hope it makes sense!

samuelsinayoko committed:
Shorten overly long lines in examples.

samuelsinayoko committed:
Shorter one line description.

samuelsinayoko committed:
Better description in the notes section.

samuelsinayoko commented:
@datapythonista Nice seeing you at the London meetup this week, thanks again for organising. Are you happy with the latest changes?

TomAugspurger committed:
Formatting [ci skip]

[ci skip]

TomAugspurger commented:
Thanks @samuelsinayoko !
